libs/wire-subsystems/src/Wire/AuthenticationSubsystem/Interpreter.hs

```diff
@@ -98,6 +105,65 @@ instance Exception PasswordResetError where
   displayException InvalidResetKey = "invalid reset key for password reset"
   displayException InProgress = "password reset already in progress"
 
+authenticateImpl ::
+  ( Member UserStore r,
+    Member (Error AuthenticationSubsystemError) r,
+    Member HashPassword r,
+    Member PasswordStore r
+  ) =>
+  UserId ->
+  PlainTextPassword6 ->
+  Sem r ()
+authenticateImpl uid plaintext = do
+  getUserAuthenticationInfo uid >>= \case
+    Nothing -> throw AuthenticationSubsystemInvalidUser
+    Just (_, Deleted) -> throw AuthenticationSubsystemInvalidUser
+    Just (_, Suspended) -> throw AuthenticationSubsystemSuspended
+    Just (_, Ephemeral) -> throw AuthenticationSubsystemEphemeral
+    Just (_, PendingInvitation) -> throw AuthenticationSubsystemPendingInvitation
+    Just (Nothing, _) -> throw AuthenticationSubsystemBadCredentials
+    Just (Just password, Active) -> do
+      case Pasword.verifyPasswordWithStatus plaintext password of
+        (False, _) -> throw AuthenticationSubsystemBadCredentials
+        (True, PasswordStatusNeedsUpdate) -> do
+          for_ (plainTextPassword8 . fromPlainTextPassword $ plaintext) (hashAndUpdatePwd uid)
+        (True, _) -> pure ()
+  where
+    hashAndUpdatePwd u pwd = do
+      hashed <- hashPassword8 pwd
+      upsertHashedPassword u hashed
+
+-- | Password reauthentication. If the account has a password, reauthentication
+-- is mandatory. If the account has no password, or is an SSO user, and no password is given,
+-- reauthentication is a no-op.
+reauthenticateImpl ::
+  ( Member (Error AuthenticationSubsystemError) r,
+    Member UserStore r,
+    Member UserSubsystem r,
+    Member (Input (Local ())) r
+  ) =>
+  UserId ->
+  Maybe (PlainTextPassword' t) ->
+  Sem r ()
+reauthenticateImpl user plaintextMaybe =
+  getUserAuthenticationInfo user >>= \case
+    Nothing -> throw AuthenticationSubsystemInvalidUser
+    Just (_, Deleted) -> throw AuthenticationSubsystemInvalidUser
+    Just (_, Suspended) -> throw AuthenticationSubsystemSuspended
+    Just (_, PendingInvitation) -> throw AuthenticationSubsystemPendingInvitation
+    Just (Nothing, _) -> throw AuthenticationSubsystemBadCredentials
+    Just (Just pw', Active) -> maybeReAuth pw'
+    Just (Just pw', Ephemeral) -> maybeReAuth pw'
+  where
+    maybeReAuth pw' = case plaintextMaybe of
+      Nothing -> do
+        local <- input
+        musr <- getLocalAccountBy NoPendingInvitations (qualifyAs local user)
```

Comment:
there is `qualifyLocal'` in `Brig.App`. i guess you could move that here?

Reply:
Would need to move it somewhere it can be accessed in multiple modules without creating reverse or cyclic dependencies. Feels like more work than we need to do right now.